### ISSUE TITLE
Vonneumannize the Transpiler

### ISSUE THREAD

**Issue by jwoehr** [qiskit-advocate/qamp-spring-23#7]
### Description

Transpilation is largely a multifaceted optimization problem of the sort that Quantum Computing proposes to handle.

In the spirit of "eating one's own cooking", we will explore ways to use quantum computing in quantum program transpilation.

There have been [suggestions of this sort](https://github.com/Qiskit/feedback/discussions/3) in years past when devices were smaller and more limited. Today, with larger devices, the problem may be more tractable.

Much of the work will still be classical (file i/o, factoring the problem, etc.). However, my take on this is that employing the device to transpile for itself is a sort of sanity check on Quantum Computing itself.

### Deliverables

A version of the transpiler that employs actual quantum computing hardware in certain instances where practical.

### Mentors details

- Mentor 1
  - Name: Jack Woehr
  - GitHub ID: @jwoehr
  - What they do: Qiskit Advocate and author

### Number of mentees

3

### Type of mentees

- Mentee 1
  - Required:
    - Quantum computing knowledge
    - Interested in QASM-level fundamentals of quantum program preparation
  - Nice to have:
    - Strong mathematical background
    - Knowledge of gate model
- Mentee 2
  - Required:
    - Quantum computing knowledge
    - Proficient in Qiskit
  - Nice to have:
    - Experience in Qiskit open source contributions
- Mentee 3
  - Required:
    - Good  written and verbal communication skills
    -  Good work habits
  - Nice to have:
    - Markdown experience
    - Presentation experience (E.g., Google Slides, PPT, or whatever)

We will want 2 hardcore developers who will code and 1 documentation expert who will maintain the project website and blog which will effectively document our project.

**Comment by jwoehr** [comment]
Note to Qiskit Team -- this project may require extended access to larger devices.

**Comment by omarcostahamido** [comment]
yes!
(sorry, just want to be in the loop 😄)

**Comment by jwoehr** [comment]
@omarcostahamido YOU ARE DRAFTED! 🥇

**Comment by gunchamalik** [comment]
@jwoehr I am interested in learning more about transpilation and can contribute to this project for the mentee 3 skills.

**Comment by jwoehr** [comment]
Interesting paper on related subject [Quantum-assisted quantum compiling](https://quantum-journal.org/papers/q-2019-05-13-140/#)

**Comment by jwoehr** [comment]
@gunchamalik hoping some more Advocates sign up soon! Then we can meet and chat about this idea.

**Comment by mpham26uchicago** [comment]
Hi @jwoehr , are you still taking Mentee 1/2? I have experience in transpilation, specifically in unitary decomposition.

**Comment by jwoehr** [comment]
Yes, @mpham26uchicago you are the first to ask about those positions in the mentorship. But it looks like you have proposed your own project, #33 !
